Add filter for specific architectures

Hyask opened the pull request:
This PR moves the `platforms` and `platforms_exe` variables to the configuration, and adds a filter in the `rustup` download mechanism to avoid mirroring all the available architectures.

This is fixes #6 .

YetiBarBar commented on src/rustup.rs:
```diff
@@ -217,6 +219,7 @@ pub fn rustup_download_list(
             .flat_map(|(_, pkg)| {
                 pkg.target
                     .into_iter()
+                    .filter(|(name, _)| { platforms.contains(&name) || platforms_exe.contains(&name) } )
```
As platform list is moved in configuration files, this list is now a user input (and not compiled in binary).

Thus, with your modification there is no runtime-check that a user-provided platform (through mirror.toml) exist. As a result, panamax will report a success even when user ask to develop that non-existent platform.

dureuill replied:
Hello (I worked with @Hyask on the code in this MR),

Should we maybe keep the list of supported platforms in the code, and check the provided input list against this list of supported platforms reporting an error in the case the input configuration is incorrect?

YetiBarBar replied:
As a simple user, I would a least expect a log message if my user input seems to be inexistent.

From my point of view (note that I'm not-related with panamax maintenance, just a single user), your code introduce a big breaking change in Panamax configuration:
- platform and platform_exe are now mandatory in the configuration files ;
- when updating panamax, admin will have to change configuration file in their repo ==> most lazy admin who don't read changelog carefully will have to deal with an unexpected error message and most probably won't think they have to review their mirror.toml file because of new mandatories fields.

==> Maybe you can make this new rustup section field optional and, if not available, fallback to the "good'old" behaviour the sysadmin is used to...

dureuill replied:
This looks like a good point (albeit a bit different from handling incorrect configuration), thank you for your input!

We will update the MR to:

- Make the setting optional, and keep the existing behavior if the option is absent (or the list empty). Whether or not we'd like to put an informational message in that case, informing that all configurations will be downloaded, depends on what the maintainer(s) think(s) is best.
- When the list is populated by the user through the configuration, check that the input platforms exist and are known of panamax, and emit an error if it not the case.

k3d3 commented:
This looks great, thank you so much @Hyask and @dureuill!

I will echo that it would be better for the platforms and platforms_exe config values to be optional, so as to not break the backwards compatibility of existing installations (even though technically Panamax doesn't have any stability promise yet). It also works as a way to say "vacuum up absolutely everything" when the platforms are not specified.

k3d3 commented:
As a somewhat unrelated side-note, I was considering changing the method used for discovering which rustup-init variants are available. It appears you can make anonymous `ls` requests to static.rust-lang.org as it runs on s3. That means if Panamax could perform the equivalent of

```
aws s3 ls --no-sign-request s3://static-rust-lang-org/rustup/dist/
```

this would allow the discovery and validation of *all* rustup architectures, not just the ones I happen to put in the code.

This isn't important for now - having a hard-coded list of architectures is fine for the time being - however I do hope to improve the architecture handling in the near future.

Hyask commented:
I've refactored a bit the code so that:
* If the configuration misses the new settings, `panamax` will log an `Info` and default to all platforms, so that this PR is fully backward-compatible with existing configurations.
* If a bad platform is specified in one of the new settings, it will log and promptly terminate with an error (currently a panic, because not handled by `main`, this can be changed).
* I've introduced a `Platforms` struct with both `unix` and `windows` fields, for easier handling of both values across all functions.
* I've renamed `platforms` to `platforms_unix` and `platforms_exe` to `platforms_windows`, to improve clarity in the settings.

Let me know what you think of it. :)

YetiBarBar commented:
Thanks for the job!

I think it's a safe behaviour.

k3d3 commented:
Looks great!

One detail I just realized is that there are some platforms that exist in the distributions that don't have a rustup-init associated with them. It does appear to be the other way around though, i.e. rustup-inits without distributions.

As far as I can tell (looking through channel-rust-stable.toml), the platforms are:
```
"aarch64-fuschia",
"aarch64-pc-windows-msvc",
"aarch64-unknown-hermit",
"aarch64-unknown-none",
"aarch64-unknown-none-softfloat",
"aarch64-unknown-redox",
"arm-unknown-linux-musleabi",
"arm-unknown-linux-musleabihf",
"armv5te-unknown-linux-gnueabi",
"armv5te-unknown-linux-musleabi",
"armv7-apple-ios",
"armv7-unknown-linux-gnueabi",
"armebv7r-none-eabi",
"armebv7r-none-eabihf",
"armv7s-apple-ios",
"asmjs-unknown-emscripten",
"i386-apple-ios",
"i586-pc-windows-msvc",
"i586-unknown-linux-gnu",
"i586-unknown-linux-musl",
"i686-unknown-freebsd",
"i686-unknown-linux-musl",
"mips64-unknown-linux-muslabi64",
"mips64el-unknown-linux-muslabi64",
"mipsisa32r6el-unknown-linux-gnu",
"mipsisa64r6-unknown-linux-gnuabi64",
"mipsisa64r6el-unknown-linux-gnuabi64",
"nvptx64-nvidia-cuda",
"riscv32i-unknown-none-elf",
"riscv32imc-unknown-none-elf",
"riscv32imac-unknown-none-elf",
"riscv32gc-unknown-linux-gnu",
"riscv64imac-unknown-none-elf",
"riscv64gc-unknown-none-elf",
"sparc64-unknown-linux-gnu",
"sparcv9-sun-solaris",
"thumbv6m-none-eabi",
"thumbv7em-none-eabi",
"thumbv7neon-linux-androideabi",
"thumbv7neon-unknown-linux-gnueabihf",
"wasm32-unknown-emscripten",
"wasm32-unknown-unknown",
"wasm32-wasi",
"x86_64-apple-ios",
"x86_64-fortanix-unknown-sgx",
"x86_64-fuschia",
"x86_64-pc-solaris",
"x86_64-rumprun-netbsd",
"x86_64-sun-solaris",
"x86_64-unknown-linux-gnux32",
"x86_64-unknown-redox",
```

All of these platforms appear to have downloads, or are at the very least referred to in the channel.toml files. Having this many more platforms than rustup-inits is a bit of a pain, but the sake of this PR, I'd recommend two more things:

1) Allow a platform to silently error out when downloading a rustup-init file (in case it's one of the above platforms).

2) Add all of these to PLATFORMS_UNIX (except for `i586-pc-windows-msvc`, to be added to PLATFORMS_WINDOWS).

In the future, I can look into something that dynamically checks the channel.toml files for validating the mirror config, or handles rustup-inits and toolchain platforms separately, but this PR will at least get some amount of filtering into the hands of people who want it.

Thank you again for the work you've done on this PR!

Hyask commented:
I pushed two more commits with your suggestions. Tell me what you think :)

k3d3 commented:
Looks good to me, thank you!
